src/net_processing.cpp

```diff
@@ -1888,16 +1888,23 @@ void static ProcessOrphanTx(CConnman* connman, std::set<uint256>& orphan_work_se
 }
 
 /**
- *  A block has been processed. If this is the first time we've seen the block,
+ *  A block has been processed. If the block was failed anti-dos / mutation checks, then
+ *  call BlockChecked() to maybe punish peer. If this is the first time we've seen the block,
  *  update the node's nLastBlockTime. Otherwise erase it from mapBlockSource.
  */
-void static BlockProcessed(CNode* pfrom, std::shared_ptr<CBlock> pblock, bool new_block)
+void static BlockProcessed(CNode* pfrom, CConnman* connman, std::shared_ptr<CBlock> pblock, BlockValidationState& state, bool new_block)
 {
-    if (new_block) {
-        pfrom->nLastBlockTime = GetTime();
-    } else {
+    if (!state.IsValid()) {
+        // The block failed anti-dos / mutation checks. Call BlockChecked() callback here.
+        // This clears the block from mapBlockSource.
+        BlockChecked(*pblock, state, connman);
+    } else if (!new_block) {
+        // Block was valid but we've seen it before. Clear it from mapBlockSource.
         LOCK(cs_main);
         ::mapBlockSource.erase(pblock->GetHash());
```

Comment:
Can't this branch combined with `BlockChecked` by passing new_block as param? Even wonder if we still need `fNewBlock` in PNB, as  `BlockChecked`  erase block from `mapBlockSource` independently of validity/novelty

Reply:
`BlockChecked()` will only be called from `BlockProcessed()` if the block failed `CheckBlock()`/`AcceptBlock()`. This else branch is catching the case where the block didn't fail, but we've seen it before.

If you think this can be cleaned up further, perhaps we can do it in a follow-up PR.